src/SteamAPI.ts

```diff
@@ -335,46 +336,50 @@ export default class SteamAPI {
 	async getGameDetails(
 		app: number | number[],
 		{ language = this.language, currency = this.currency, filters = [] } = {}
-	): Promise<{ [key: string]: any }> {
+	): Promise<{ [key: number]: GameDetails }> {
 		assertApp(app);
 
 		const isArr = Array.isArray(app);
-		const key = `${app}-${currency}-${language}`;
-
-		// For now we're not touching the cache if an array of apps is passed
-		// TODO: maybe cache apps individually if an array is passed?
-		if (!isArr) {
-			const cached = this.gameDetailCache?.get(key);
-			if (cached) return cached;
+		const appIDs = isArr ? app : [app];
+
+		const cached = appIDs.map(a => this.gameDetailCache?.get(`${a}-${currency}-${language}`)).filter(g => g !== undefined);
+		const remainingAppIDs = appIDs.filter((_, i) => cached[i] === undefined);
+
+		// If some appIDs were missing from the cache, fetch from Steam and update the cache
+		if (remainingAppIDs.length !== 0) {
+			const details = await this
+				.get('/appdetails', {
+					appids: remainingAppIDs.join(','),
+					cc: currency,
+					l: language,
+					filters: isArr && appIDs.length > 1 ? 'price_overview' : filters.join(','),
+				}, this.baseStore)
+				.then(json => {
+					if (json === null) throw new Error('Failed to find app ID');
+
+					const filtered: { [key: string]: any } = {};
+					for (const [k, v] of Object.entries(json))
+						if ((v as any).success) {
+							const d = (v as any).data;
+							// Convert empty arrays to empty objects for consistency
+							filtered[k] = Array.isArray(d) && d.length === 0 ? {} : d;
+						}
+
+					if (Object.keys(filtered).length === 0) throw new Error('Failed to find app ID');
+
+					return filtered;
+				});
+
+			for (const [appID, data] of Object.entries(details)) {
+				const game = new GameDetails(data);
+				this.gameDetailCache?.set(`${appID}-${currency}-${language}`, game);
+				cached.push(game);
+			}
 		}
 
-		const details = await this
-			.get('/appdetails', {
-				appids: isArr ? app.join(',') : app,
-				cc: currency,
-				l: language,
-				filters: isArr && app.length > 1 ? 'price_overview' : filters.join(','),
-			}, this.baseStore)
-			.then(json => {
-				if (json === null) throw new Error('Failed to find app ID');
-
-				// TODO: make a class
-				const filtered: { [key: string]: any } = {};
-				for (const [k, v] of Object.entries(json))
-					if ((v as any).success) {
-						const d = (v as any).data;
-						// Convert empty arrays to empty objects for consistency
-						filtered[k] = Array.isArray(d) && d.length === 0 ? {} : d;
-					}
-
-				if (Object.keys(filtered).length === 0) throw new Error('Failed to find app ID');
-
-				return isArr ? filtered : filtered[app];
```

Comment:
This line in particular made me think this function was returning either a single details directly or an object of ID -> details mappings.